Create Bot Instances during initial bot join

timothyb89 opened the pull request:
Part of https://github.com/gravitational/teleport/issues/41456

This creates new instances for bots when they initially join the cluster, and persists instance IDs in new certificate fields on join and during renewal.

Note that this does not yet handle instance reuse for non-token join methods.

Additionally, bot instance creation is locked behind a `BOT_INSTANCE_EXPERIMENT` flag; it must be set to `1` to enable creation.

No changelog entry yet as this is still largely invisible to users (without the flag, at least).

timothyb89 committed:
Create Bot Instances during initial bot join

This creates new instances for bots when they initially join the
cluster, and persists instance IDs in new certificate fields on join
and during renewal.

Note that this does not yet handle instance reuse for non-token join
methods.

Additionally, bot instance creation is locked behind a
`BOT_INSTANCE_EXPERIMENT` flag; it must be set to `1` to enable
creation.

timothyb89 commented on lib/auth/join.go:
```diff
@@ -380,6 +368,22 @@ func (a *Server) generateCertsBot(
 		},
 	}
+
+	fingerprint := sha256.Sum256(req.PublicTLSKey)
```
I know we specified a "sha256 fingerprint" in the RFD but I'm not actually entirely clear on exactly what that is. I've just hashed and hex encoded it here.

strideynet replied:
TBH - The hash/fingerprint mostly existed for the purposes of identifying a BotInstance when the initial plan wasn't to use UUIDs - we could probably just drop this field now.

If you want to keep it, there's not a whole lot of agreement on fingerprinting/hashing public keys. In other places of Teleport, e.g the TPM authentication, I achieved this by hashing the PKIX encoded public key. However, it looks like our Join endpoints support the public key being submitted in either PKIX or PKCS1 encoding - if we are going to have a hash, we should de-encode from PKIX or PKCS1 and re-encode always into one specific encoding before hashing to ensure that the same public key receives the same fingerprint regardless of encoding.

timothyb89 replied:
Fair enough, I'll drop the field. I think it'd be a tricky identifier regardless - easier than the public key, sure, but still a difficult thing for end users to look at and compare vs the UUID.

timothyb89 commented on lib/auth/join.go:
```diff
@@ -389,7 +393,32 @@ func (a *Server) generateCertsBot(
 			log.WithError(err).Warn("Unable to encode join attributes for audit event.")
 		}
+
+		auth.Metadata, err = structpb.NewStruct(attributes)
```
Also not certain of what metadata means per the RFD - I assume it's this (join method claims)?

strideynet replied:
Yup - this is what I meant 😅

timothyb89 commented on lib/auth/auth_with_roles.go:
```diff
@@ -3251,6 +3251,12 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
+		// Always pass through a bot instance ID if available. Note that this
+		// method is only used for bot identity renewals and is not responsible
+		// for issuing new instance IDs; see `generateInitialBotCerts()`
+		// TODO: need to update bot instance with new authentication + generation counter
```
We'll need to construct a new authentication record here, and I think it creates a couple weird cases I'm not certain of:
* should we add an auth record just for renewals, or also for impersonated certs? (I assume the former)
* what is the join token value for this path? Should it just be `""`?
* what will metadata be? empty?

Re-authenticated delegated joins should all go through the `generateInitialBotCerts()` route rather than here, so this _should_ just be `token` method renewals, assuming we don't include impersonated certs.

strideynet replied:
> should we add an auth record just for renewals, or also for impersonated certs? (I assume the former)

Definitely just for renewals and not for impersonated certs. It'll be much noisier if we do impersonated certs.

> what is the join token value for this path? Should it just be ""?

Yeah - I'd think so. 

> what will metadata be? empty?

Yeah - empty - at least in terms of identity metadata. We could include information like the generation counter etc but I'm not convinced it really adds anything compared to the rich metadata we get from somewhere like GitHub.

timothyb89 replied:
Works for me. We do have the generation counter in the auth status proto already, so I can't think of anything else we'd want to add.

timothyb89 commented on lib/auth/bot.go:
```diff
@@ -216,6 +234,28 @@ func (a *Server) generateInitialBotCerts(ctx context.Context, botName, username,
+			BotName:    botName,
+			InstanceId: uuid.String(),
+
+			// TODO: set TTL? should we rely on the known expiration instead?
+			// (auth may overwrite the value now or later)
```
Not 100% sure, but I may regret introducing a TTL field to the bot spec given every renewal provides an expiry. We'll need to pass that along somehow, or come up with a sensible value for this field another way.

strideynet replied:
It feels to me like we could just directly leverage the metadata.expiry field, I'm not too sure having a separate TTL field has any positive impact - no harm in dropping it from the protos in one of these PRs.

timothyb89 replied:
Yeah, I'll go ahead and make that change. I figure we're safe to make breaking changes to the protos still, since they should never have been exercised on any real clusters.

timothyb89 committed:
Proto cleanup, and update bot auth records on cert renewal

This makes various (admittedly breaking) protobuf changes, including
removing the TTL field (calculating resource expiry based on cert
requests), removing public key fingerprints, and changing the data
type of the generation counter to match the preexisting internal
datatype. These changes _should_ be safe as no consumers of the proto
API currently exist.

Additionally, this also updates bot authentications on renewal.

timothyb89 commented on api/proto/teleport/machineid/v1/bot_instance.proto:
```diff
@@ -101,11 +97,9 @@ message BotInstanceStatusAuthentication {
   // method, this counter is checked during renewal and the Bot is locked out if
   // the counter in the certificate does not match the counter of the last
   // authentication.
-  int32 generation = 5;
```
This is a breaking proto change, I _think_ we can get away with this because there's no real consumers of this API yet, but if we're worried about it we can either keep this as an int32 or swap this for a v2 field and do it "properly".

I can also mark the removed fields as reserved, in that case.

strideynet replied:
Unfortunately - the linter can't be overriden 😓  You'll probably need to just roll to a new field name and number

timothyb89 replied:
Ah, always forget about the linter. It's harmless enough to mark the removed fields as `reserved`, but I think I'll drop the type change. Some napkin math suggests the smaller type really won't cause a problem:
```
In [1]: 2147483647 / (3 * 24 * 365)
Out[1]: 81715.51168188736
```

I figure ~82 thousand years of continuous renewals ought to be enough for anybody 🤷 

(Oddly, the proto language guide it links to explicitly says `int32` and `uint64` are compatible, but the linter complains about it anyway, oh well.)

strideynet replied:
Hah, I think 82 thousand ought to be enough for anyone 😉

strideynet commented on lib/auth/bot.go:
```diff
@@ -168,14 +174,86 @@ func (a *Server) validateGenerationLabel(ctx context.Context, username string, c
 	return nil
 }
 
+func (a *ServerWithRoles) updateBotAuthentications(ctx context.Context, req *certRequest) error {
```
I assume we need to account for there not being an ident.BotInstanceID if this is a cert from pre-BotInstances that is renewing.

timothyb89 replied:
Yep, good call, I've added a case to handle that.

timothyb89 committed:
Fix proto lints

timothyb89 commented on lib/auth/machineid/machineidv1/bot_instance_service.go:
```diff
@@ -33,6 +34,18 @@ import (
+	// ExpiryMargin is the duration added to bot instance expiration times to
+	// ensure the instance remains accessible until shortly after the last
+	// issued certificate expires.
+	ExpiryMargin = time.Minute * 5
```
I'm curious if there's any thoughts on the expiry margin. I imagine we want some nonzero buffer to account for any clock weirdness and so renewals don't fail because the bot instance somehow got pruned before a cert expires. I've just guessed at 5 minutes being good enough for this purpose.

That said, is there any value in keeping bot instances around longer? Do they provide worthwhile informational value? Or is it more valuable for a bot instance to expire shortly after a bot can no longer renew, so `tctl get bot_instance/foo` shows an approximately correct list of active bots?

strideynet replied:
So from a technical angle, I think about 5 minutes makes sense to account for any "weirdness".

From a product angle, I think that for "historical data" we already have an audit log and this is probably what we ought to leverage for "I want to see info about things that happened in the past" rather than "show me the active state of my cluster". We probably ought to make sure long-term that anything that's accessible via the BotInstance is also accessible via the audit log. I do think it's quite nice w/ the current design that `tctl get bot_instance` shows anything thats currently capable of authenticating.

timothyb89 replied:
Makes sense to me, I'll keep it at 5 minutes (and make a note to ensure audit events include an actionable amount of info on the bot instance, since they'll be acting as the historical record)

timothyb89 committed:
Fix misleading doc comment in the bot instance experiment

timothyb89 committed:
Create bot instances for old bots on join; other fixes

This now creates bot instances for bots whose certs are missing the
BotInstanceID field. Additionally, it fixes two related bugs:
expiration dates are extended on renewal, the generated UUID
is properly appended to certs on initial join, and instances are
only created or updated when the experiment is enabled.

timothyb89 committed:
Add a minimal test for bot instance creation on initial join

timothyb89 committed:
Validate bot instance state in generation counter checks

timothyb89 committed:
Remove outdated TODO comment and fix test lints

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bot-instance-creation

timothyb89 committed:
Add an expiration change check to the generation test

strideynet reviewed (approved):
Looks good !
